gcore/gdaldriver.cpp

```diff
@@ -323,8 +324,20 @@ GDALDatasetH CPL_DLL CPL_STDCALL GDALCreate(GDALDriverH hDriver,
 {
     VALIDATE_POINTER1(hDriver, "GDALCreate", nullptr);
 
-    return GDALDriver::FromHandle(hDriver)->Create(
+    GDALDatasetH hDS = GDALDriver::FromHandle(hDriver)->Create(
         pszFilename, nXSize, nYSize, nBands, eBandType, papszOptions);
+
+#ifdef OGRAPISPY_ENABLED
+    if (nBands < 1)
+    {
+        OGRAPISpyCreateDataSource(reinterpret_cast<OGRSFDriverH>(hDriver),
```

Comment:
this cast and the reinterpret_cast<OGRDataSourceH>(hDS) ones are not correct. Here this should work because the implementation of OGRAPISpyCreateDataSource actually recast them to GDALDriverH and GDALDatasetH, but if it would case OGRSFDriverH (which here is a GDALDriver*) to the corresponding OGRSFDriver* that would be technically incorrect (although OGRSFDriver extends GDALDriver without any member)
We can just avoid those casts by changing the signature and implementation of OGRAPISpyCreateDataSource to accept a GDALDriverH and GDALDatasetH.

Reply:
Thanks, fixed with https://github.com/OSGeo/gdal/pull/9591/commits/8ebfe91eec7417bc551c1b0cb842ff47e5fd0023